instance/models/utils.py

```diff
@@ -423,9 +433,48 @@ class ConsulAgent(object):
     managing prefixes, and reduces the call size to the main needed things with a
     possibility to expand it for more advanced queries.
     """
-    def __init__(self, prefix=''):
+
+    # We use the `consul lock` CLI to acquire locks on the server side.
+    # Those use a "magic flag" to help indicate conflicts with semaphores.
+    # We want to compete with those locks though, so we use this flag too.
+    # Comes from https://github.com/hashicorp/consul/blob/master/api/lock.go#L35-L38
+    MAGIC_LOCK_FLAG = 0x2ddccbc058a50c18
+
+    def __init__(self, prefix='', lock_wait=100, lock_wait_sleep=1):
         self._client = consul.Consul()
         self.prefix = prefix
+        self.session_id = None
+        self.lock_wait = lock_wait
+        self.lock_wait_sleep = lock_wait_sleep
+
+    def __enter__(self):
+        # Any `put` operation using a session with "delete" behavior is ephemeral.
+        # So if we create the lock below with this session, when the session is
+        # invalidated (TTL completed, session was destroyed, etc.) the lock key
+        # is also deleted. However there is a default lock delay of 15 seconds, so
+        # new locks have to wait for that timer even if the lock key is physically gone.
+        # See https://www.consul.io/docs/internals/sessions.html for details.
+        self.session_id = self._client.session.create(behavior="delete")
```

Comment:
shouldn't node be provided as well?

Reply:
@dupainaulevain https://python-consul.readthedocs.io/en/latest/#consul.base.Consul.Session.create - it'll use the current node.